### PR TITLE
[Merged by Bors] - fix: Fix `FundamentalGroupoid` being reducible.

### PR DESCRIPTION
**Pull request by adamtopaz**
Instead it is now the opposite -- a structure with a single field.

[zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Groupoid.2EtoCategory/near/400820106)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean [hunk 1]
```diff
@@ -42,7 +42,9 @@ class SimplyConnectedSpace (X : Type*) [TopologicalSpace X] : Prop where
 theorem simply_connected_iff_unique_homotopic (X : Type*) [TopologicalSpace X] :
     SimplyConnectedSpace X ↔
       Nonempty X ∧ ∀ x y : X, Nonempty (Unique (Path.Homotopic.Quotient x y)) := by
-  rw [simply_connected_def, equiv_punit_iff_unique]; rfl
+  rw [simply_connected_def, equiv_punit_iff_unique]
+  refine ⟨fun ⟨⟨x⟩,h⟩ => ⟨⟨x.as⟩, fun x y => h _ _⟩,
+    fun ⟨⟨x⟩,h⟩ => ⟨⟨⟨x⟩⟩, fun x y => h _ _⟩ ⟩
```
Comment (eric-wieser) [inline Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean:47]: Is there a missing lemma here? Maybe `Nonempty (FundamentalGroupoid X) \iff Nonempty X` (and similarly for `Subsingleton` and `Nontrivial`)

File: Mathlib/AlgebraicTopology/FundamentalGroupoid/Basic.lean [hunk 2]
```diff
@@ -268,33 +268,34 @@ end Path
 
 /-- The fundamental groupoid of a space `X` is defined to be a type synonym for `X`, and we
 subsequently put a `CategoryTheory.Groupoid` structure on it. -/
-def FundamentalGroupoid (X : Type u) := X
+@[ext]
+structure FundamentalGroupoid (X : Type u) where
+  as : X
```
Comment (eric-wieser) [inline Mathlib/AlgebraicTopology/FundamentalGroupoid/Basic.lean:273]: I suspect the linter will ask for this anyway, but; can you add a docstring to `as`, preferably one that makes clear the intended mnemonic?

File: Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean:47]: (the trick is to build the equiv and then all these results are there by dot notation)

Comment (adamtopaz) [reply Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean:47]: Probably. I didn't think too much when I made these changes :) I'll fix this soon.

**Comment by alreadydone** [comment]
I noticed this a while ago and have [another design](https://github.com/leanprover-community/mathlib4/pull/7886/files#diff-2729c9e3f2b1add9310a9490643d4afb6edd64d8f99902e32ae4cff562c74fe4R246-R251) to fix the problem without changing FundamentalGroupoid to a one-field structure, but I haven't fixed all errors. The ideas is that Lean can't quite identify the coercion of `{ α := FundamentalGroupoid X } : Grpd` to a Type* with `FundamentalGroupoid X` reducibly, and the former is used a lot, so we may just define `FundamentalGroupoid X` to be `{ α := X, str := ... } : Grpd` and Lean won't confuse it with `X` itself.

Did you find this because you want to do something with fundamental groupoids, or by `import Mathlib`? I was doing something about van Kampen and want to define a [functor](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/n.20composable.20morphisms/near/399407412) from the unit interval (with the category structure from its order) to a fundamental groupoid but noticed that the unit interval actually got the fundamental groupoid structure as well.

File: Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean (hunk 1 above)
Comment (adamtopaz) [reply Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean]: I added `equiv` and some related lemmas and instances. I don't want this PR to expand too far beyond the original fix, so I would prefer to leave anything further which is not directly related to a future PR.

File: Mathlib/AlgebraicTopology/FundamentalGroupoid/Product.lean [hunk 3]
```diff
@@ -55,17 +55,18 @@ theorem proj_map (i : I) (x₀ x₁ : πₓ (TopCat.of (∀ i, X i))) (p : x₀
 #align fundamental_groupoid_functor.proj_map FundamentalGroupoidFunctor.proj_map
 
 -- Porting note: losing the instance with a concrete category again
-instance : (i : I) → TopologicalSpace (πₓ (X i)).α := fun i => TopCat.topologicalSpace_coe (X i)
+--instance : (i : I) → TopologicalSpace (πₓ (X i)).α :=
```
Comment (eric-wieser) [inline Mathlib/AlgebraicTopology/FundamentalGroupoid/Product.lean:58]: Should this be deleted? Or do we need TO DO here?

Comment (alreadydone) [reply Mathlib/AlgebraicTopology/FundamentalGroupoid/Product.lean:58]: Yeah I think it should be. The fundamental groupoid shouldn't get a topological structure, just like the topological space shouldn't get a category structure. Previously they got mixed up due to `reducible`.

**Comment by eric-wieser** [comment]
bors d=@alreadydone

**Review by alreadydone — APPROVED** [review]
Thanks! This approach seems to cause the least breakage, as all fixes are straightforward, and I think it will be good enough for my purpose.

bors r+

**Pull request closed without merging**